src/MIDebugEngine/Engine.Impl/DebugUnixChildProcess.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using MICore;
+using System.Diagnostics;
+using Microsoft.DebugEngineHost;
+
+namespace Microsoft.MIDebugEngine
+{
+    public interface ProcessSequence
+    {
+        Task Enable(bool enable);
+        /// <summary>
+        /// Handle a stopping event as part of a sequence of debugger operations
+        /// </summary>
+        /// <param name="debugEvent"></param>
+        /// <returns>true if stopping event was consumed (process must be running), false otherwise (indicting the debugger should process the event)</returns>
+        Task<bool> Stopped(Results debugEvent);
+        /// <summary>
+        /// Handle breakpoint created events
+        /// </summary>
+        /// <param name="debugEvent"></param>
+        /// <returns>true if event was consumed, false otherwise (indicting the debugger should process the event)</returns>
+        bool BreakpointCreated(Results debugEvent);
+    }
+
+    class DebugUnixChild : ProcessSequence
+    {
+        private string _forkBp;
+        private string _vforkBp;
+
+        private enum State
+        {
+            Init,
+            Enabled,
+            AtFork,
+            AtVfork,
+            AtSignal,
+            AtExec
+        }
+        private State _state;
+        private DebuggedProcess _process;
+        private int _newpid;
+        private LaunchOptions _launchOptions;
+        private string _mainBreak;
+
+        public DebugUnixChild(DebuggedProcess process, LaunchOptions launchOptions)
+        {
+            _process = process;
+            _state = State.Init;
+            _launchOptions = launchOptions;
+        }
+
+        private async Task ProcessChild()
+        {
+            Debug.Assert(_newpid != 0, "Child process id not found.");
+            if (_newpid != 0)
+            {
+                int inf = _process.InferiorByPid(_newpid);
+                if (inf != 0)
+                {
+                    await _process.ConsoleCmdAsync("inferior " + inf.ToString());
+                    if (!string.IsNullOrEmpty(_mainBreak))
+                    {
+                        await _process.MICommandFactory.BreakDelete(_mainBreak);
+                        _mainBreak = null;
+                    }
+                    _state = State.AtSignal;
+                    await _process.MICommandFactory.Signal("SIGSTOP");  // stop the child
```

Comment:
It does not wait for the signal to be received. The behavior of signal is identical to issuing a continue, except that is continues after delivering the named signal.